multi-layer-perceptron/jax_test.py

```diff
@@ -0,0 +1,308 @@
+        """
+        w_key, b_key = random.split(key)
+        
+        return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))
+
+    def __init_layers(self, key=random.PRNGKey(0) ):
+        """
+        Initialize all layers for a fully-connected neural network.
+        """
+        sizes = self.layer_sizes
+        keys = random.split(key, len(sizes))
+
+        return [self.__init_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]
+
+def relu(x):
+    """
+    Rectified Linear Operation.
+    """
+    return jnp.maximum(0, x)
+
+def one_hot_encoder(x, k, dtype=jnp.float32):
+    """
+    Create a one-hot encoding of x of size k.
+    """
+    return jnp.array(x[:, None] == jnp.arange(k), dtype)
+
+def predict(params, input):
+    """
+    Run a single prediction on the input. 
+
+    Parameters: 
+        input (array): A n-dimensional array, typically a 1D or 2D array. For images for instance this is a 2D array. 
+    """
+
+    activations = input
+    for w, b in params[:-1]:
+        outputs = jnp.dot(w, activations) + b
+        activations = relu(outputs)
+    
+    final_w, final_b = params[-1]
+    logits = jnp.dot(final_w, activations) + final_b
+    
+    # Compute log softmax loss
+    loss = logits - logsumexp(logits)
+
+    return loss
+
+# Temp, move this
+batched_predict = vmap(predict, in_axes=(None, 0))
+
+def loss(params, inputs, targets):
+    """
+    Computes negative log likehood loss. 
+
+    Each prediction of batched_predict retuns the log softmax loss.
+    Computing the mean and returning the mean is this form of loss. 
+    """
+    preds = batched_predict(params, inputs)
+    loss = -jnp.mean(preds * targets)
+
+    return loss
+
+@jit
+def update(params, x, y, learning_rate):
+    """
+    Perform backpropagation using stochastic gradient descent.
+    """
+    grads = grad(loss)(params, x, y)
+
+    return [(w - learning_rate * dw, b - learning_rate * db) for (w, b), (dw, db) in zip(params, grads)]
+
+def accuracy(params, inputs, targets):
+    """
+    Test accuracy 
+    """
+
+    target_class = jnp.argmax(targets, axis=1)
+    predicted_class = jnp.argmax(batched_predict(params, inputs), axis=1)
+    
+    return jnp.mean(predicted_class == target_class)
+
+def train(params, ds, train_images, train_labels, test_images, test_labels, num_labels, num_pixels, learning_rate, epochs=10, debug=False):
+    print("\nTraining Model...")
+
+    training_metrics = {}
+    ts = time.time()
+
+    epoch_times = []
+
+    for epoch in tqdm(range(epochs)):
+
+        start_time = time.time()
+
+        # Main training loop
+        for x, y in ds:
+            x = jnp.reshape(x, (len(x), num_pixels))
+            y = one_hot_encoder(y, num_labels)
+            params = update(params, x, y, learning_rate)
+        
+        epoch_time = time.time() - start_time
+        epoch_times.append(epoch_time)
+
+        train_acc = accuracy(params, train_images, train_labels)
```

Comment:
For consistency with other implementation, exclude this call to accuracy here since it'll do another whole loop through the training data to do inference. Most of the others do the accuracy prediction at the same time as processing the batch during the update step and having another run through the training data will add time to each epoch.

Reply:
Moved so that accuracy is only called is debug is enabled, so no accuarcy by default.